src/Kurrent.Client/Core/Serialization/MessageTypeMapper.cs

```diff
@@ -0,0 +1,76 @@
+using System.Collections.Concurrent;
+
+namespace Kurrent.Client.Tests.Streams.Serialization;
+
+public interface IMessageTypeRegistry {
+	void    Register(Type messageType, string messageTypeName);
+	string? GetTypeName(Type messageType);
+	string  GetOrAddTypeName(Type clrType, Func<Type, string> getTypeName);
+	Type?   GetClrType(string messageTypeName);
+	Type?   GetOrAddClrType(string messageTypeName, Func<string, Type?> getClrType);
+}
+
+public class MessageTypeRegistry : IMessageTypeRegistry {
+	readonly ConcurrentDictionary<string, Type?> _typeMap     = new();
```

Comment:
Do we need concurrent dictionaries here? Type mappings are normally done at startup and not at runtime.